File: packages/storage/ios/Plugin/FirebaseStorage.swift
import Foundation
import FirebaseCore
import FirebaseStorage

@objc public class FirebaseStorage: NSObject {
    private let plugin: FirebaseStoragePlugin

    init(plugin: FirebaseStoragePlugin) {
        self.plugin = plugin
        super.init()
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }
    }

    @objc public func deleteFile(_ options: DeleteFileOptions, completion: @escaping (Error?) -> Void) {
        let path = options.getPath()

        let storageRef = Storage.storage().reference(withPath: path)
        storageRef.delete { error in
            if let error = error {
                completion(error)
            } else {
                completion(nil)
            }
        }
    }

    @objc public func getDownloadUrl(_ options: GetDownloadUrlOptions, completion: @escaping (Result?, Error?) -> Void) {
        let path = options.getPath()

        let storageRef = Storage.storage().reference(withPath: path)
        storageRef.downloadURL { url, error in
            if let error = error {
                completion(nil, error)
            } else {
                let result = GetDownloadUrlResult(url: url!)
                completion(result, nil)
            }
        }
    }

    @objc public func getMetadata(_ options: GetMetadataOptions, completion: @escaping (Result?, Error?) -> Void) {
        let path = options.getPath()

        let storageRef = Storage.storage().reference(withPath: path)
        storageRef.getMetadata { metadata, error in
            if let error = error {
                completion(nil, error)
            } else {
                let result = GetMetadataResult(metadata: metadata!)
                completion(result, nil)
            }
        }
    }

    @objc public func listFiles(_ options: ListFilesOptions, completion: @escaping (Result?, Error?) -> Void) {
        let path = options.getPath()
        let maxResults = options.getMaxResults()
        let pageToken = options.getPageToken()

        let storageRef = Storage.storage().reference(withPath: path)
        if let pageToken = pageToken {
            storageRef.list(maxResults: maxResults, pageToken: pageToken) { result, error in
                if let error = error {
                    completion(nil, error)
                } else {
                    let result = ListFilesResult(result: result!)
                    completion(result, nil)
                }
            }
        } else {
            storageRef.list(maxResults: maxResults) { result, error in
                if let error = error {
                    completion(nil, error)
                } else {
                    let result = ListFilesResult(result: result!)
                    completion(result, nil)
                }
            }
        }
    }

    @objc public func updateMetadata(_ options: UpdateMetadataOptions, completion: @escaping (Error?) -> Void) {
        let path = options.getPath()
        let metadata = options.getMetadata()

        let storageRef = Storage.storage().reference(withPath: path)
        storageRef.updateMetadata(metadata) { _, error in
            if let error = error {
                completion(error)
            } else {
                completion(nil)
            }
        }
    }

    @objc public func uploadFile(_ options: UploadFileOptions, completion: @escaping (Result?, Error?, Bool) -> Void) {
        let path = options.getPath()
        let uri = options.getUri()
        let metadata = options.getMetadata()

        let storageRef = Storage.storage().reference(withPath: path)
        let uploadTask: StorageUploadTask
        if let metadata = metadata {
            uploadTask = storageRef.putFile(from: uri, metadata: metadata)
        } else {
            uploadTask = storageRef.putFile(from: uri)
        }
        uploadTask.observe(.progress) { snapshot in
            let result = UploadFileCallbackEvent(snapshot: snapshot)
            completion(result, nil, false)
        }
        uploadTask.observe(.success) { snapshot in
            let result = UploadFileCallbackEvent(snapshot: snapshot)
            completion(result, nil, true)
        }
        uploadTask.observe(.failure) { snapshot in
            completion(nil, snapshot.error, true)
        }
    }
    
    @objc func useEmulator(_ host: String, _ port: Int) {
        Storage.storage().useEmulator(withHost: host, port: port)
    }
}

File: packages/storage/android/src/main/java/app/capgo/capacitor/firebase/storage/FirebaseStorage.java
package app.capgo.capacitor.firebase.storage;

import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import app.capgo.capacitor.firebase.storage.classes.events.UploadFileCallbackEvent;
import app.capgo.capacitor.firebase.storage.classes.options.DeleteFileOptions;
import app.capgo.capacitor.firebase.storage.classes.options.GetDownloadUrlOptions;
import app.capgo.capacitor.firebase.storage.classes.options.GetMetadataOptions;
import app.capgo.capacitor.firebase.storage.classes.options.ListFilesOptions;
import app.capgo.capacitor.firebase.storage.classes.options.UpdateMetadataOptions;
import app.capgo.capacitor.firebase.storage.classes.options.UploadFileOptions;
import app.capgo.capacitor.firebase.storage.classes.results.GetDownloadUrlResult;
import app.capgo.capacitor.firebase.storage.classes.results.GetMetadataResult;
import app.capgo.capacitor.firebase.storage.classes.results.ListFilesResult;
import app.capgo.capacitor.firebase.storage.enums.UploadFileState;
import app.capgo.capacitor.firebase.storage.interfaces.EmptyResultCallback;
import app.capgo.capacitor.firebase.storage.interfaces.NonEmptyEventCallback;
import app.capgo.capacitor.firebase.storage.interfaces.NonEmptyResultCallback;
import com.google.android.gms.tasks.Task;
import com.google.firebase.storage.ListResult;
import com.google.firebase.storage.StorageMetadata;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

public class FirebaseStorage {

    private FirebaseStoragePlugin plugin;

    public FirebaseStorage(FirebaseStoragePlugin plugin) {
        this.plugin = plugin;
    }

    public void deleteFile(@NonNull DeleteFileOptions options, @NonNull EmptyResultCallback callback) {
        String path = options.getPath();

        StorageReference storageReference = getFirebaseStorageInstance().getReference(path);
        storageReference
            .delete()
            .addOnSuccessListener(aVoid -> callback.success())
            .addOnFailureListener(exception -> callback.error(exception));
    }

    public void getDownloadUrl(@NonNull GetDownloadUrlOptions options, @NonNull NonEmptyResultCallback callback) {
        String path = options.getPath();

        StorageReference storageReference = getFirebaseStorageInstance().getReference(path);
        storageReference
            .getDownloadUrl()
            .addOnSuccessListener(uri -> {
                GetDownloadUrlResult result = new GetDownloadUrlResult(uri);
                callback.success(result);
            })
            .addOnFailureListener(exception -> callback.error(exception));
    }

    public void getMetadata(@NonNull GetMetadataOptions options, @NonNull NonEmptyResultCallback callback) {
        String path = options.getPath();

        StorageReference storageReference = getFirebaseStorageInstance().getReference(path);
        storageReference
            .getMetadata()
            .addOnSuccessListener(storageMetadata -> {
                GetMetadataResult result = new GetMetadataResult(storageMetadata);
                callback.success(result);
            })
            .addOnFailureListener(exception -> callback.error(exception));
    }

    public void listFiles(@NonNull ListFilesOptions options, @NonNull NonEmptyResultCallback callback) {
        String path = options.getPath();
        int maxResults = options.getMaxResults();
        String pageToken = options.getPageToken();

        StorageReference storageReference = getFirebaseStorageInstance().getReference(path);
        Task<ListResult> task;
        if (pageToken == null) {
            task = storageReference.list(maxResults);
        } else {
            task = storageReference.list(maxResults, pageToken);
        }
        task
            .addOnSuccessListener(listResult -> {
                ListFilesResult result = new ListFilesResult(listResult);
                callback.success(result);
            })
            .addOnFailureListener(exception -> callback.error(exception));
    }

    public void updateMetadata(@NonNull UpdateMetadataOptions options, @NonNull EmptyResultCallback callback) {
        String path = options.getPath();
        StorageMetadata metadata = options.getMetadata();

        StorageReference storageReference = getFirebaseStorageInstance().getReference(path);
        storageReference
            .updateMetadata(metadata)
            .addOnSuccessListener(aVoid -> callback.success())
            .addOnFailureListener(exception -> callback.error(exception));
    }

    public void uploadFile(@NonNull UploadFileOptions options, @NonNull NonEmptyEventCallback callback) {
        String path = options.getPath();
        Uri uri = options.getUri();
        @Nullable
        StorageMetadata metadata = options.getMetadata();

        StorageReference storageReference = getFirebaseStorageInstance().getReference(path);
        UploadTask uploadTask;
        if (metadata == null) {
            uploadTask = storageReference.putFile(uri);
        } else {
            uploadTask = storageReference.putFile(uri, metadata);
        }
        uploadTask
            .addOnProgressListener(taskSnapshot -> {
                UploadFileCallbackEvent result = new UploadFileCallbackEvent(taskSnapshot, UploadFileState.RUNNING);
                callback.success(result);
            })
            .addOnSuccessListener(taskSnapshot -> {
                UploadFileCallbackEvent result = new UploadFileCallbackEvent(taskSnapshot, UploadFileState.SUCCESS);
                callback.success(result);
                callback.release();
            })
            .addOnFailureListener(exception -> {
                callback.error(exception);
                callback.release();
            });
    }

    public void useEmulator(@NonNull String host, int port) {
        getFirebaseStorageInstance().useEmulator(host, port);
    }

    private com.google.firebase.storage.FirebaseStorage getFirebaseStorageInstance() {
        return com.google.firebase.storage.FirebaseStorage.getInstance();
    }
}
